src/dpg_slam/dpg_slam.cc

```diff
@@ -676,25 +737,65 @@ namespace dpg_slam {
+	 std::vector<PointIdInfo> removed_points_at_current_node;
+	 bool commitChanges = false;
+         
+	 for (const auto &occ_grid_entry : current_node_occ_grid.getGridInfo()) {
+             CellStatus local_submap_occupancy = local_submap_occ_grid.getCellStatus(occ_grid_entry.first);
+             if ((occ_grid_entry.second == OCCUPIED) && (local_submap_occupancy == FREE)) {
+                   std::vector<PointIdInfo> new_added = current_node_occ_grid.getPointsInOccCell(occ_grid_entry.first);
+                   added_points_at_current_node.insert(added_points_at_current_node.end(), new_added.begin(), new_added.end());
+		   ROS_INFO_STREAM("Added");
+            	} else if ((occ_grid_entry.second == FREE) && (local_submap_occupancy == OCCUPIED)) {
+                   std::vector<PointIdInfo> new_removed = local_submap_occ_grid.getPointsInOccCell(occ_grid_entry.first);
+                   removed_points_at_current_node.insert(removed_points_at_current_node.end(), new_removed.begin(), new_removed.end());
+		   ROS_INFO_STREAM("Removed");
+            	}
             }
-        }
+ 	 
+	 // if newly detected changes at node are non zero, check the score.
+	 if (added_points_at_current_node.size()+removed_points_at_current_node.size() > 0) {
+	 	commitChanges = computeBinScoreAndCommitLabelsForNode(added_points_at_current_node,
+									   removed_points_at_current_node);
+	 }
+	 if (!commitChanges) {
+		return;
+	 }
+
+	 // commit changes if we get here.
+	 added_points.insert(added_points.end(), added_points_at_current_node.begin(), added_points_at_current_node.end());
+	 removed_points.insert(removed_points.end(), removed_points_at_current_node.begin(), removed_points_at_current_node.end());
     }
 
-    std::vector<DpgNode> DpgSLAM::updateActiveAndDynamicMaps(const std::vector<DpgNode> &nodes, const std::vector<dpgMapPoint> &removedPoints){
-
-        std::vector<DpgNode> inactiveNodes;
-        //TODO: Implement Alg. 2 from the paper.
-
-
-        return inactiveNodes;
+    bool DpgSLAM::computeBinScoreAndCommitLabelsForNode(const std::vector<PointIdInfo> &added_points,
+		    					const std::vector<PointIdInfo> &removed_points) {
+
+         uint16_t totalBins = dpg_parameters_.num_bins_for_change_detection_;
+	 double change_threshold = dpg_parameters_.delta_change_threshold_;
+	 std::vector<uint16_t> changedBins;
+	 bool commitChanges = false;
+	 double current_changed_ratio = 0.0;
+	 std::vector<PointIdInfo> changedPoints = added_points;
+	 changedPoints.insert(changedPoints.end(), removed_points.begin(), removed_points.end());
+
+	 // I know this looks messy but could not think of a better way to get total number of points at a node.
+	 uint64_t node_num = changedPoints.back().node_num_;
+	 uint32_t totalPointsAtNode = dpg_nodes_[node_num].getMeasurement().getMeasurements().size();
+
+	 for (uint32_t i=0; i< changedPoints.size(); i++) {
+	 	uint32_t binNumber = round(totalBins*changedPoints[i].point_num_in_node_/totalPointsAtNode);
```

Comment:
Does this actually get the bin number? I don't think this quite works because, while the point_num_in_node is a scaled version of the angle for points in the node, that doesn't hold for the removed points, which will be from different nodes. I think you'll have to do a similar thing to project the point into polar coordinates and sort by angle as is done in the sector deactivation. You'll probably also need to worry about the angle_min/angle_max, but maybe that's sort of covered since this has already filtered out those points to those in the field of view.

Reply:
If I understand you correctly, this method wouldn't be an issue for added points since they are all from the same node but not for removed because they are coming from different nodes and I was using totalPointsAtNode from a fixed node. ...?
I'm still slightly confused about the issue, I'll try to do it as you suggested and we can discuss tomorrow if I'm not getting it right.